### ISSUE TITLE
new code cli's installation workflow/"version" subcommand is confusing/maybe buggy

### ISSUE THREAD

**Issue by leifwalsh** [microsoft/vscode#192225]
I've tried downloading the new code CLI from https://code.visualstudio.com/insiders/ (the x64 CLI build for Linux: `code-insiders 1.82.0 (commit f1302be1e67e3af5fbeb8bbb2ea784de7bc96150)`) and I'm trying to understand how it can be used to manage and maybe install different versions of VS Code, ultimately to package up something that can run `code serve-web` for us.

I have some possible bugs with `code version use`, and some questions:
1. There is a typo in the word "requested" here https://github.com/microsoft/vscode/blob/main/cli/src/util/errors.rs#L274
2. I can't get `code version use` to do anything, I think maybe arg parsing is broken?

   ```
   ❯ ./code-insiders version use
   error: the following required arguments were not provided:
     <stable | insiders | x.y.z | path>
   
   Usage: code-insiders version use <stable | insiders | x.y.z | path>
   
   For more information, try '--help'.
   
   ❯ ./code-insiders version use stable
   [2023-09-05 18:06:22] error The reqested version is invalid, expected one of 'stable', 'insiders', version number (x.y.z), or absolute path.
   
   ❯ ./code-insiders version use insiders
   [2023-09-05 18:06:27] error The reqested version is invalid, expected one of 'stable', 'insiders', version number (x.y.z), or absolute path.
   
   ❯ ./code-insiders version use ~/.vscode-server-insiders/bin/9800cf6dd6bf4634889d60720ef46a400f3a7298
   No installation of Visual Studio Code /home/leif/.vscode-server-insiders/bin/9800cf6dd6bf4634889d60720ef46a400f3a7298 was found.
   
   If you already installed Visual Studio Code and we didn't detect it, run `code version use /home/leif/.vscode-server-insiders/bin/9800cf6dd6bf4634889d60720ef46a400f3a7298 --install-dir /path/to/installation`
   
   ❯ ./code-insiders version use --install-dir ~/.vscode-server-insiders/bin/9800cf6dd6bf4634889d60720ef46a400f3a7298 insiders
   [2023-09-05 18:07:13] error The reqested version is invalid, expected one of 'stable', 'insiders', version number (x.y.z), or absolute path.
   ```
3. It seems like `code ext install` doesn't work until it thinks an installation is available:

   ```
   ❯ ./code-insiders ext install path/to/internal-extension.vsix
   No installation of Visual Studio Code insider was found.
   Install it from your system's package manager or https://code.visualstudio.com/insiders, restart your shell, and try again.
   
   If you already installed Visual Studio Code and we didn't detect it, run `code version use insider --install-dir /path/to/installation`
   ```

   But, fine. So I need to get the code cli to know about or create an installation it likes.
4. I tried just running `./code-insiders serve-web` and it was able to download and install a copy of VS Code on the first request. So clearly it can manage installations. But is there a way to explicitly ask the code CLI to do the install up front (before the first request)? Ideally I'd like to set things up to install a copy of the server, install a few required extensions, and then run serve-web so that it's ready to go from the start. Also it would be nice to specify whether we want to install stable or insiders - given the help text for `code version use` it sounds like you are trying to do that, I just can't figure out the CLI yet.

   When we tried the preview of `code-server serve-local`, that did a download and install on startup, before starting the server, which was closer to that. The old `code-server` CLI also would agree to install extensions before installing the server/whole VS Code application.

I guess overall, is what I'm trying to do here within the scope of things you're aiming for the CLI to be able to do? Am I hitting bugs or just holding it wrong? Is there anything I can read about how this is intended to work yet? Would it be helpful for me to explain in more detail what our use cases are?

**Comment by connor4312** [comment]
`version use` and `ext` apply only to desktop installations of VS Code. (Version previously was able to manage multiple qualities, but we removed that capability; it looks like there are still some traces of that in the help text.) That's also why it failed to attach to the path in your ~/.vscode-server-insiders -- it didn't recognize it as a valid VS Code (desktop) installation.

There is no equivalent "ext install" command for serve-web at present.

**Comment by sliekens** [comment]
@connor4312 so what is the intended usage of `version use`?

I can only instruct it to use a preinstalled instance like

```
code use version stable c:\development\code\vscode-win32-x64-1.82.2
```

I thought it worked more like NVM

```
code version use insiders
[2023-09-23 15:14:41] error The reqested version is invalid, expected one of 'stable', 'insiders', version number (x.y.z), or absolute path.

code version use 1.82.0
[2023-09-23 15:15:03] error The reqested version is invalid, expected one of 'stable', 'insiders', version number (x.y.z), or absolute path.
```

**Comment by leifwalsh** [comment]
Ok. If it's now intended that the server will just download new updates if available when connecting, that seems fine. Makes our proxy management a little more complicated, but it was already going to be that way. 

Follow-up questions:
1. Without a desktop install, it seems like most of the other features of the CLI don't really work. Is there a way to get it to interact with the running server when that exists? For example, using `code path/to/file` to open a file in an existing window. Or does that still work by having the server inject another `PATH` component into terminals you run while inside it?
2. Should I file a separate ticket about adding the ability to install extensions into the serve-web version? Or is there some other way to do that? (Can I just unpack a VSIX into the right location?)

**Comment by leifwalsh** [comment]
1. It seems like once you're connected, inside terminals there your path still gets adjusted so that `code` is the server's code cli, rather than the rust one. So `code path/to/file` works, but not if your shell is from outside VS Code. I think that's the same as what `code-server` got you.
2. Ping?

**Comment by SoCuul** [comment]
I'm having the same problem on macOS, running stable. For some reason replacing `stable` with `oss` fixed it.
```sh
$ code version use stable
[2023-11-25 16:27:08] error The reqested version is invalid, expected one of 'stable', 'insiders', version number (x.y.z), or absolute path.
```

**Comment by scottcwang** [comment]
I dug into @leifwalsh's second question (how to install extensions via CLI to a `serve-web` instance). Here's what I found:
- All `serve-web` (and `tunnel`) instances use the path `${HOME}/.vscode-server` to store extensions and user data. This means that you can plop an install of full vscode onto a build machine and run `code --extensions-dir /tmp/path --install-extension abc.def --install-extension ghi.jkl`. This will populate `/tmp/path` with the extensions, which you can copy to `${HOME}/.vscode-server` on the host where `serve-web` will be running.
```bash
vscode_data_dir=$(mktemp --directory)
mkdir ${vscode_data_dir}/extensions

echo ${extensions_to_install} |
    xargs printf "--install-extension '%s' " |
    xargs --no-run-if-empty \
        code --extensions-dir ${vscode_data_dir}/extensions

# If you have user settings that should go in too, use this:
echo ${user_settings} > ${vscode_data_dir}/data/Machine/settings.json

cp ${vscode_data_dir} ${HOME}/.vscode-server
```
- The `serve-web` command waits for the HTTP endpoint to be requested, then downloads the vscode server to the directory `${HOME}/.vscode/cli/serve-web/<commit hash>`. In there, `bin/code-server` has an `--install-extension` option. In other words, the following commands will install the extensions, but they will work only after someone has requested the HTTP endpoint and the vscode server is downloaded:
```bash
# It probably isn't too important to get the latest commit hash, but if needed, lru.json provides the order of commits if needed
most_recent_server_path="${HOME}/.vscode/cli/serve-web/$(
    jq --raw-output --monochrome-output ".[0]" ${HOME}/.vscode/cli/serve-web/lru.json
)"

echo ${extensions_to_install} |
    xargs printf "--install-extension '%s' " |
    xargs --no-run-if-empty \
        ${most_recent_server_path}/bin/code-server
```
- (You didn't ask, but) The `tunnel` command is similar, but `most_recent_server_path` is `${HOME}/.vscode/cli/servers/Stable-<commit hash>/server` instead.
- `serve-web` downloads the vscode server `.tar.gz` from `https://update.code.visualstudio.com/commit:<commit hash>/server-linux-x64-web/stable/`, where the hash of the latest commit comes from `https://update.code.visualstudio.com/api/latest/linux-x64/stable/`. (`tunnel` downloads a different `.tar.gz` from `https://update.code.visualstudio.com/commit:<commit hash>/server-linux-x64/stable/`.) Aside from decompressing the `.tar.gz` to the server location in "${HOME}/.vscode/cli/serve-web/<commit hash>`, I don't think the CLI does anything else; so if you want to predownload the vscode server before the first time the HTTP endpoint is requested, you could try these URLs and see what happens. The source code for the CLI's update manager is [here](https://github.com/microsoft/vscode/blob/main/cli/src/update_service.rs#L108-L171).

**Comment by leifwalsh** [comment]
Awesome, thanks. This is a bit manual, but I can make it work!
